### PR TITLE
ZEN-31219: fix error handling in maintenence cycle

### PR DESCRIPTION
**Pull request by lundybernard**
Fixes: ZEN-31219

### CODE REVIEW

File: Products/ZenCollector/daemon.py [hunk 1]
```diff
@@ -832,53 +832,51 @@ def _startMaintenance(self, ignored=None):
...
+        try:
+            if not self.options.cycle:
+                ret = "No maintenance required"
+            elif getattr(self.preferences, 'pauseUnreachableDevices', True):
+                # TODO: handle different types of device issues, such as WMI issues
+                issues = yield self.getDevicePingIssues()
+                ret = yield self._processDeviceIssues(issues)
+        except Exception as err:
+            self.log.exception('failure in _maintenanceCycle')
+            raise err
```
Comment (jpeacock-zenoss) [inline Products/ZenCollector/daemon.py]: Just need `raise` here, I think.  `raise err` deletes the traceback.

File: Products/ZenCollector/daemon.py [hunk 2]
```diff
@@ -832,53 +832,51 @@ def _startMaintenance(self, ignored=None):
...
-                    d = defer.succeed(None)
-            else:
-                d = defer.succeed("No maintenance required")
-            return d
+        defer.returnValue(ret)
 
-        d = _maintenance()
-        return d
+    @defer.inlineCallbacks
+    def _processDeviceIssues(self, result):
```
Comment (jpeacock-zenoss) [inline Products/ZenCollector/daemon.py]: There's no I/O occurring in this method, so there's no need for it to use inlineCallbacks.

Comment (lundybernard) [reply Products/ZenCollector/daemon.py]: Fixed

File: Products/ZenCollector/daemon.py (hunk 1 above)
Comment (lundybernard) [reply Products/ZenCollector/daemon.py]: Fixed

File: Products/ZenCollector/daemon.py [hunk 3]
```diff
@@ -832,53 +832,50 @@ def _startMaintenance(self, ignored=None):
...
+        for devId in clearedDevices:
+            self.log.debug("Resuming tasks for device %s", devId)
+            self._scheduler.resumeTasksForConfig(devId)
+
+        self._unresponsiveDevices = newUnresponsiveDevices
+        for devId in self._unresponsiveDevices:
+            self.log.debug("Pausing tasks for device %s", devId)
+            self._scheduler.pauseTasksForConfig(devId)
+
+        defer.returnValue(result)
```
Comment (jpeacock-zenoss) [inline Products/ZenCollector/daemon.py]: Need a regular return here.

File: Products/ZenCollector/daemon.py [hunk 4]
```diff
@@ -832,53 +832,50 @@ def _startMaintenance(self, ignored=None):
...
+        except Exception:
+            self.log.exception('failure in _maintenanceCycle')
+            raise
+
+        defer.returnValue(ret)
+
+    def _processDeviceIssues(self, result):
+        self.log.debug("deviceIssues=%r", result)
+        if result is None:
+            defer.returnValue(result)  # exception or some other problem
```
[the remark this answers is not in the document]
Comment (lundybernard) [reply Products/ZenCollector/daemon.py]: refactored

**Pull request merged**